packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.test.ts

```diff
@@ -0,0 +1,645 @@
+      ],
+    },
+  ],
+};
+
+describe('insight user actions preprocessing', () => {
+  describe('#mapAndSortActionsByMostRecent', () => {
+    it('should properly map and sort the raw user actions into UserAction objects', async () => {
+      const expectedMappedAndSortedActions = [
+        {
+          actionType: 'VIEW',
+          timestamp: '1648826040000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {},
+        },
+        {
+          actionType: 'CLICK',
+          timestamp: '1648825140000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {title: 'title', uriHash: 'TtnKwc0Lo2GY9WAi'},
+        },
+        {
+          actionType: 'CUSTOM',
+          timestamp: '1648822620000',
+          eventData: {type: 'useless_event', value: ''},
+          searchHub: 'community-support',
+          document: {},
+        },
+        {
+          actionType: 'CLICK',
+          timestamp: '1648822500000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {title: 'title', uriHash: 'KXñi9EWk38wnb1tt'},
+        },
+        {
+          actionType: 'SEARCH',
+          timestamp: '1648822380000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {},
+          query: '',
+        },
+        {
+          actionType: 'VIEW',
+          timestamp: '1648744500000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {},
+        },
+        {
+          actionType: 'CLICK',
+          timestamp: '1648744320000',
+          eventData: {},
+          searchHub: 'in-product-help',
+          document: {title: 'title', uriHash: 'caCgiG2JPzjZfS7G'},
+        },
+        {
+          actionType: 'CUSTOM',
+          timestamp: '1648744290000',
+          eventData: {
+            type: 'smartSnippetSuggestions',
+            value: 'expandSmartSnippetSuggestion',
+          },
+          searchHub: 'in-product-help',
+          document: {},
+        },
+      ];
+      const mockRawActions = [...fakeActions].slice(0, 8);
+      const mappedAndSortedActions =
+        mapAndSortActionsByMostRecent(mockRawActions);
+
+      expect(mappedAndSortedActions).toEqual(expectedMappedAndSortedActions);
+      expect(Number(mappedAndSortedActions[0].timestamp)).toBeGreaterThan(
+        Number(mappedAndSortedActions[1].timestamp)
+      );
+    });
+
+    it('should log a warning in the console if a rawUserAction cannot be parsed', () => {
+      const invalidRawUserAction = {
+        name: 'Custom',
+        value: '{ "key": "value",,,, }',
+        time: '2023-10-01T12:00:00Z',
+      };
+
+      const consoleWarnSpy = jest
+        .spyOn(console, 'warn')
+        .mockImplementation(
+          () =>
+            'Some user actions could not be parsed. Please check the raw user actions data.'
+        );
+
+      mapAndSortActionsByMostRecent([invalidRawUserAction]);
```

Comment:
We need also check the return value of `mapAndSortActionsByMostRecent`, even when an invalid action is passed, we wanna make sure `mapAndSortActionsByMostRecent` returns an array as a proof that `mapAndSortActionsByMostRecent` doesn't break in this situation.